projects/bootloader/src/config.c

```diff
@@ -0,0 +1,87 @@
+#include "config.h"
+#include "crc32.h"
+#include "flash.h"
+#include "log.h"
+#include "persist.h"
+#include "soft_timer.h"
+
+static PersistStorage s_config_1_persist = { 0 };
+static BootloaderConfig s_config_1_blob = { 0 };
+
+static PersistStorage s_config_2_persist = { 0 };
+static BootloaderConfig s_config_2_blob = { 0 };
+
+#define BOOTLOADER_CONFIG_PAGE_1_FLASH_PAGE (FLASH_ADDR_TO_PAGE(BOOTLOADER_CONFIG_PAGE_1_START))
+#define BOOTLOADER_CONFIG_PAGE_2_FLASH_PAGE (FLASH_ADDR_TO_PAGE(BOOTLOADER_CONFIG_PAGE_2_START))
+
+StatusCode config_init(void) {
```

Comment:
The safety features going on in this function are rather involved, so you should have a big comment explaining what it's doing (and why), and small comments each time it does something non-self-explanatory (like loading from flash or restoring a corrupt page).